Reorganize source structure to prevent source shadowing of installed package

carterbox opened the pull request.

carterbox committed:
BLD: Replace VERSION with setuptools_scm

Instead of manually managing the VERSION file, setuptools_scm
will automatically create a version number based on git tags.
CMake interrogates git separately and does not include the git
hash on the version number because it only allows numbers.

I chose setuptools_scm instead of versioneer because it doesn't
require adding any additional files the the repo. Instead all of
the logic is contained within the scm package which is a
dependency which is automatically installed by setup.py at
install time or it can be pre-installed in the environment.

carterbox committed:
REF: Reorganize files so source doesn't overshadow installed

By moving the python module to /src, devs have the choice of
testing against the source code or installed code by running the
tests from either at / or inside /tests.

carterbox committed:
REF: Move tests down one directory because if __init__.py

test is actually a python module because they have an
__init__.py. This means that the modules in src are also imported
and the source will always overshadow the installed tomopy.
Moving the tests module down one directory solves this problem.

carterbox committed:
REF: Adjust CMake to new file structure

Appended /src to file paths and put if(NOT SKBUILD) around copy
operations that are not necessary if building without an IDE.

carterbox committed:
REF: Change directories for coverage and tests

carterbox committed:
DOC: Correct spelling languae -> language

jrmadsen reviewed (changes requested):
In other projects I have found it most common that `include` and `src` are typically at the same level so since we are moving include down a folder, maybe make the directory `source`.

Also, a lot of things were deleted from the main CMakeLists.txt... Particularly `project(tomopy LANGUAGES C VERSION ${tomopy_VERSION})` that line should absolutely not be removed. 

Also, many of the things you deleted in other CMake files are there so that one can actually work with tomopy in an IDE. I suggest that we keep them and not do them if we are building from python, e.g. `if(NOT SKBUILD)`

Unless I missed you putting it somewhere else:
```
include(MacroUtilities)
include(Options)
include(Compilers)
include(GNUInstallDirs)
include(ProjectSettings)
include(BuildSettings)
include(Packages)
include(ClangFormat)
```
is kind of important... that is where we set compiler flags, find the packages we need to link to, etc.

carterbox commented on src/CMakeLists.txt:
```diff
@@ -1,3 +1,34 @@
+include(BuildSettings)
+include(Packages)
+include(ClangFormat)
+
```
I moved the `project` statement and the CMake module includes here: from `/` to `/src`. I did this because the directory of the `project` statement sets `${PROJECT_SOURCE_DIR}`. I thought it would be better to move the project statement than update all lines that contain `${PROJECT_SOURCE_DIR}`. The include statements are also moved because I observed that at least one of them needs to be declared after the project is declared otherwise certain variables are not set.

Was this not a good idea? Does moving the project declaration also mess up IDE builds?

jrmadsen replied:
No you should keep the project statement in the top CMakeLists.txt

carterbox replied:
What if I replace `${PROJECT_SOURCE_DIR}` with `${CMAKE_CURRENT_LIST_DIR}` inside `/src/CMakeLists.txt`?

carterbox commented on src/CMakeLists.txt:
```diff
@@ -39,40 +70,6 @@ macro(CONFIG_INSTALL RELATIVE_PATH)
-execute_process(COMMAND ${CMAKE_COMMAND} -E copy_if_different
-    ${PYSETUP} ${PROJECT_BINARY_DIR}/
-    WORKING_DIRECTORY ${PROJECT_BINARY_DIR})
-
```
These copy operations on lines 42...76 are the ones which you want to keep for IDE compatibility correct? I can gate them with `if(NOT SKBUILD)`.

Do lines 89...92 also need remain unaltered for IDEs?

carterbox commented on pyctest_tomopy.py:
```diff
@@ -220,7 +220,7 @@ def run_pyctest():
         pyctest.BUILD_NAME = pyctest.BUILD_NAME.replace("  ", " ")
     # how to build the code
-    pyctest.BUILD_COMMAND = "{} setup.py install".format(
+    pyctest.BUILD_COMMAND = "{} setup.py develop".format(
```
Notice that I also updated this line. I assumed that we wanted run tests on the source. Otherwise, `coverage` does not work because it does not know where the installed package is located.

jrmadsen replied:
Hmm... interesting side-effect. Do the scripts in benchmarking work? Bc if not, they are probably executing from `/benchmarking` and the `WORKING_DIRECTORY` needs to be changed for those tests

carterbox replied:
From my limited understanding, I believe the other benchmarks are unaffected. I ran both `/pyctest_tomopy.py` and `/benchmarking/pyctest_tomopy_phantom.py` to check, and I think they ran correctly.

In theory, this pull shouldn't affect them. The pull only changes where TomoPy is installed such that `coverage` can monitor the code while it's running. `coverage` expects to monitor `/src` because that is what `/.coveragerc` says.

I think the difference in behaviour with this pull is the following:

*Before* - Benchmarks would run on the installed TomoPy and `coverage` would run on the source TomoPy because `libtomopy.so` was always copied there during the build.

*Now* - Benchmarks and `coverage` all run on the same code because `pyctest_tomopy.py` uses `develop` instead of `install`. `libtomopy.so` is not always copied to the source during the build.

carterbox replied:
I think I resolved this problem in https://github.com/jrmadsen/tomopy/pull/6/commits/d7095e0720d82833fa008065e4144a473025ae1b. Thus, this build command is no longer changed.

jrmadsen commented:
Yea these changes are causing a problem for notifying the users about the configuration because of CMake scoping:

<img width="622" alt="screen shot 2019-02-12 at 4 39 01 pm" src="https://user-images.githubusercontent.com/6001865/52677955-d3a8ff80-2ee4-11e9-8335-87471f6998a0.png">

`cd /Users/jrmadsen/devel/c++/NERSC/build-test/src && /opt/local/bin/gcc -DNDEBUG -DTOMOPY_USE_MKL -Dlibtomopy_EXPORTS -I/Users/jrmadsen/devel/c++/NERSC/tomopy-cmake-carterbox/src/include  -isystem /opt/conda/envs/tomopy/include  -O3 -DNDEBUG -fPIC   -std=gnu11 -W -Wall -Wextra -Wno-unused-parameter -Wunused-but-set-parameter -Wno-unused-variable -fPIC -std=c11 -fopenmp-simd -o CMakeFiles/libtomopy.dir/libtomopy/c/remove_ring.c.o   -c /Users/jrmadsen/devel/c++/NERSC/tomopy-cmake-carterbox/src/libtomopy/c/remove_ring.c`

The `tomopy_C_FLAGS` is set but according to the log, they are not.

carterbox commented:
This line is called twice in the top level CMakeList.txt. Is that necessary?

```CMAKE
set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake/Modules ${CMAKE_MODULE_PATH})
```

carterbox committed:
TST: Use setup.cfg not .coveragerc to specify covered package

Using this method instead of relative paths means that
python-coverage will be able to find the installed tomopy
and we don't need to use python setup.py develop inside the
pyctest_tomopy.py

jrmadsen commented:
I am still working through all of this -- I have a massive amount of changes in the gpu-devel branch because of the C++ support so it is a bit difficult to get everything synced, especially with everything moving around so much

carterbox commented:
Thanks for the update! Let me know if I can help.

carterbox commented:
I've already merged these changes into tomopy/1.5.x.
